flink-libraries/flink-table/src/main/java/org/apache/flink/table/annotation/ExternalCatalogCompatible.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.annotation;
+
+import org.apache.flink.annotation.Public;
+import org.apache.flink.table.catalog.TableSourceConverter;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * A tableSource with this annotation represents it is compatible with external catalog, that is,
+ * an instance of this tableSource can be converted to or converted from external catalog table
+ * instance.
+ * The annotation contains the following information:
+ * <ul>
+ * <li> external catalog table type name for this kind of tableSource </li>
+ * <li> external catalog table <-> tableSource converter class </li>
+ * </ul>
+ */
+@Documented
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Public
+public @interface ExternalCatalogCompatible {
```

Comment:
I thought about the version and agree that we do not necessarily need a dedicated version field.
However, I think we need a bit more "safety" in the translation process since the converters will rely on certain properties to be set and the link between catalog table and converter is just a string. IMO, we need some kind of verification that the properties of a `ExternalCatalogTable` are compatible with the requirements of a converter.

What do you think about the following approach:
- We extend the `TableSourceConverter` interface by a few methods to describe its requirements:
```
def requiredProperties: Array[String]
def optionalProperites: Array[String]
def getPropertyDescription(p: String): String
```
- we remove the `ExternalCatalogCompatible` annotation
- we add a `TableType` annotation which carries a String for the table type name and which is annotated to the `TableSourceConverter` (not to the TableSource!). 

By doing this, we scan for converters instead of table sources. IMO, this has the benefit that the required properties of a table type (i.e., the identifier string) are tied to the converter, which knows how to handle the ExternalTable and convert it into a TableSource. The type of the TableSource is not really relevant. It also means that there is one place to go when looking for the requirements of a table type (the converter).

What do you think @beyond1920 and @KurtYoung ?